src/sources/socket/tcp.rs

```diff
@@ -77,34 +84,58 @@ pub struct RawTcpSource {
     pub config: TcpConfig,
 }
 
+pub struct RawTcpSourceContext {
+    pub source_type: Bytes,
+    pub host_key: String,
+    pub decode: Box<dyn Fn(Bytes) -> crate::Result<Value> + Send + Sync>,
+}
+
 impl TcpSource for RawTcpSource {
+    type Context = RawTcpSourceContext;
```

Comment:
Why is this the only type that implements `Context`/`build_context`? I'm a little fuzzy on why we need them at all, but it seems like a weird interaction between the new decoders and `build_event`. It'd be nice if we could unify those two instead of adding a third mechanism.

Reply:
I introduced that so that implementors of `TcpSource` can have a context without runtime overhead, and preemptively fail building the context.

E.g. previously in `RawTcpSource` there was some logic to determine the host key which didn't depend on runtime configuration and I didn't want to build the `Decode` transform for each event anew.

Alternatively, `TcpSource::run` can have a new argument for `Decode`. However, that doesn't apply to all `TcpSource`s.

The other implementors of `TcpSource` just don't happen to have any context (yet).